### PR TITLE
Make it POSIX

### PR DESCRIPTION
**Pull request by hlangeveld**
Two major updates:

This removes any '...shisms' not compatible with POSIX.  We're not going back to pure Bourne shell, which I consider too primitive.  See the individual deltas for what changed.  Two major things were:

* Arithmetic notation, which I fixed with a hack: prefix every `(( ... ))` expression with `: $`.

* Simple, single bracket tests:  `[ ... ]` instead of `[[ ... ]]`

The second change is to throw all code into functions, this makes testing of functions simple, as we can just source the shpec script and test functions separately.  It also simplifies recursive calls to 
shpec. However, recursive calls do not update the assertion and failure counters, as its internals
run in a subshell, to protect the environment.

Performance: I had a simple set of 37 assertions which run in ~12 ms on my laptop.
There's little difference between bash, ksh and dash, except that bash takes an extra 8-10 ms to start itself.

I'm still looking for the shell equivalent of python's `if __name__ == '__main__':`
Until then I need to trigger on the value of `$0`.

### CODE REVIEW

**Comment by AdrieanKhisbe** [comment]
That's a greate PR! :)
Just add a look about it, sound great! 
(was also thinking we should make shpec a function for more flexibility and perf)

Will try to test it out in the weekend.

Could you explain what the hack about arithmetic expression does?

(about the if main. i think the way to go in bash is with the `$0`)

**Comment by erichs** [comment]
@hlangeveld: so awesome!

**Comment by rylnd** [comment]
:heart_eyes: This looks amazing. It's quite a significant PR, so please give me some time to 1) understand it and 2) provide some feedback.

Thanks so much @hlangeveld !

**Comment by hlangeveld** [comment]
On 04/10/15 09:26, Adriean Khisbe wrote:
> Could you explain what the hack about arithmetic expression does?
>


I think I can.

tldr; The hack just takes an assignment masquerading as an expression, but discards the string as
an argument of `: `, a shell builtin equivalent to `true`.

Detailed explanation:

(Note: In this example, I'll use the markdown convention of using `...` and the four space indent
to highlight code samples.)


Let's take the very simple expression:

     : $((a=1))

It consists of two parts:  `:` and `$(( a=1 ))` (arithmetic expressions are liberal about white space).

The shell builtin `:` is an actual command, accepts arguments, discards them, and returns status 0.
It's therefore an equivalent of `true`.

As `:` is a builtin, any side effects during its parameter substitution (which happens before it is executed)
are *preserved*

The arithmetic expression actually contains an assignment, and like many modern language, the value
of the assignment is equal to the value of the left hand side after assignment. Therefore,

     : $(( a=1 ))

is equivalent to:

     : 1

with a side-effect of assigning `1` to variable `a`

Note, it is not *fully* equivalent, because an arithmetic statement will return
true (exit status 0) if its numeric value equals non-zero.  Otherwise, it will be false (1).

     :  any parameter expansion not resulting in errors

will always return true (exit status 0).


A precedent can be found in very early *sh documentation;

     : ${variable:=value}

I believe this was already valid syntax for the Bourne shell, together with:

*  ${var:+replacement}	# if $var is set, expand to "replacement".

*  ${var:-default}      # expand to "default" if $var is not set (or empty)

*  ${var:?error} 	# send message "error" to stderr if $var not set, and exit current (sub)shell


I hope I made myself clear.

Cheers,
Henk

**Comment by AdrieanKhisbe** [comment]
@hlangeveld thanks for the explanation :).
Didn't know at all about the `:`.
(shell features are infinite ^^)

**Comment by hlangeveld** [comment]
On 04/11/15 11:39, Adriean Khisbe wrote:
> @hlangeveld <https://github.com/hlangeveld> thanks for the explanation :).
> Didn't know at all about the |:|.
> (shell features are infinite ^^)

A bit of personal history: Back in '85, I had an instructor who
challenged us to write shell scripts with a minimal amount
of calls to outside code (i.e., no exec() system calls).

Somehow that stuck with me. Since reading Michael Feathers' '... Legacy Code'
I found inspiration for making shells scripts (components) test-ready.

Even if you don't actually write a test (but you should), your code will
be better if you write it with tests in mind.

Cheers,
Henk

File: bin/shpec [hunk 1]
```diff
@@ -1,41 +1,51 @@
-#!/usr/bin/env bash
-VERSION=0.1.2
-examples=0
-failures=0
-test_indent=0
-red="\033[0;31m"
-green="\033[0;32m"
-norm="\033[0m"
+#
```
Comment (rylnd) [inline bin/shpec]: Does this line have a purpose?

File: bin/shpec [hunk 2]
```diff
@@ -1,41 +1,51 @@
...
+: Use one of: bash, dash, or ksh93
```
Comment (rylnd) [inline bin/shpec]: This is simply a comment, correct? Is there a reason it begins with a `:` and not a `#`?

File: bin/shpec [hunk 3]
```diff
@@ -1,41 +1,51 @@
...
 
 describe() {
-  ((test_indent += 1))
+  : $((test_indent += 1))
   iecho "$1"
 }
 
 end() {
-  ((test_indent -= 1))
+  : $((test_indent -=1 ))
```
Comment (rylnd) [inline bin/shpec]: Whitespace here is inconsistent with the near-identical line 29 above.

File: bin/shpec [hunk 4]
```diff
@@ -44,97 +54,138 @@ stub_command() {
...
-
-    time for file in $files; do
-      . "$file"
-    done
 
-    [ $failures -eq 0 ] && color=$green || color=$red
-    echo -e "${color}${examples} examples, ${failures} failures${norm}"
+shpec() {
+  (
+  VERSION=0.1.2posix
```
Comment (rylnd) [inline bin/shpec]: Let's keep the version numbers compliant with [semver](http://semver.org/), i.e. just MAJOR.MINOR.PATCH. I think a minor version bump is appropriate here.

File: bin/shpec [hunk 5]
```diff
@@ -44,97 +54,138 @@ stub_command() {
...
+      matcher_files="$(
+          find "$SHPEC_ROOT/matchers" -name '*.sh' 2>/dev/null
+      )"
+
+      for matcher_file in $matcher_files; do
+        . "$matcher_file"
+      done
+
+      if [ $# -gt 0 ] ; then
+        : \$1="'$1'" \${#1}="'${#1}'"
```
Comment (rylnd) [inline bin/shpec]: Can you explain this line?

File: bin/shpec [hunk 6]
```diff
@@ -44,97 +54,138 @@ stub_command() {
...
+        files="${@}"
+      else
+        files=$(
+          find $SHPEC_ROOT -name '*_shpec.sh'
+        )
+      fi
+
+      for file in $files; do
+        echo >& 2 "$file"
```
Comment (rylnd) [inline bin/shpec]: Please remove this line.

File: bin/shpec [hunk 7]
```diff
@@ -44,97 +54,138 @@ stub_command() {
...
+
+      for matcher_file in $matcher_files; do
+        . "$matcher_file"
+      done
+
+      if [ $# -gt 0 ] ; then
+        : \$1="'$1'" \${#1}="'${#1}'"
+        files="${@}"
+      else
+        files=$(
```
Comment (rylnd) [inline bin/shpec]: Please quote this in the same manner as `matcher_files` above.

File: bin/shpec [hunk 8]
```diff
@@ -44,97 +54,138 @@ stub_command() {
...
+          find $SHPEC_ROOT -name '*_shpec.sh'
+        )
+      fi
+
+      for file in $files; do
+        echo >& 2 "$file"
+        . "$file" || exit
```
Comment (rylnd) [inline bin/shpec]: Can you explain the intent behind the `|| exit`?

File: bin/shpec [hunk 9]
```diff
@@ -44,97 +54,138 @@ stub_command() {
...
+      done
+
+      [ $failures -eq 0 ] && color=$green || color=$red
+      echoe "${color}${examples} examples, ${failures} failures${norm}"
+
+      times
```
Comment (rylnd) [inline bin/shpec]: Didn't even know about `times`, thanks!

File: shpec/shpec_shpec.sh [hunk 10]
```diff
@@ -35,7 +35,8 @@ describe "shpec"
 
   describe "equality matcher"
     it "handles newlines properly"
-      string_with_newline_char="new\nline"
+      string_with_newline_char="new
```
Comment (rylnd) [inline shpec/shpec_shpec.sh]: Was this intentionally changed? This test was added specifically for #23.

File: shpec/shpec_shpec.sh [hunk 11]
```diff
@@ -64,16 +65,25 @@ line'
 
   describe "passing through to the test builtin"
     it "asserts an arbitrary algebraic test"
-      assert test "[[ 5 -lt 10 ]]"
+      assert test "[ 5 -lt 10 ]"
     end
   end
 
   describe "stubbing commands"
+    # only bash lets us redefine 'exit', so use 'false'
```
Comment (rylnd) [inline shpec/shpec_shpec.sh]: ce61f728abd138ae1b8e61a57fba048e4b84b15b contains the reasoning behind this test. My preference is to leave that kind of thing in git, and not in the code itself.

Additionally, I would say that this initial test is unnecessary; the following two tests cover both the stubbed functionality and the original functionality. If any part of that stub/unstub process fails, one of these commands will indicate so.

File: bin/shpec [hunk 12]
```diff
@@ -44,97 +54,138 @@ stub_command() {
...
 
-    [ $failures -eq 0 ]
+(
+  _progname=shpec
+  _pathname=$( command -v $0 )
+  _cmdname=${_pathname##*/}
+  _main=shpec
 
-esac
+  case $_progname in (${_cmdname%.sh}) $_main "$@";; esac
```
Comment (rylnd) [inline bin/shpec]: Would you mind explaining exactly how this works?

**Comment by rylnd** [comment]
@hlangeveld all in all, this is wonderful! I've added a few comments above. Once those are resolved, we just need to update the changelog, bump the version, and merge!

**Comment by rylnd** [comment]
@hlangeveld now that `shpec` is a function defined as a subshell command, can you speak to whether this affects any existing issues (for better or worse), particularly #16 or #30?

File: bin/shpec (hunk 2 above)
Comment (AdrieanKhisbe) [inline bin/shpec]: @hlangeveld A specific reason you left off `zsh`or you just did not precise it?

File: bin/shpec (hunk 4 above)
Comment (AdrieanKhisbe) [inline bin/shpec]: I concur, the introduction of the function deserve a `0.2`:)

File: bin/shpec (hunk 5 above)
Comment (AdrieanKhisbe) [inline bin/shpec]: Which line do you refer? (since we get a ten line diff?)

File: bin/shpec (hunk 8 above)
Comment (hlangeveld) [inline bin/shpec]: On 04/13/15 04:07, Ryland Herrick wrote:

>> +
>> +      for file in $files; do
>> +        echo >& 2 "$file"
>> +        . "$file" || exit
>
> Can you explain the intent behind the ||| exit|?

Hmm... The idea was catching errors in sourcing $file.
The kind of errors where it doesn't make sense to
continue - serious syntax trouble, etc.

However, this would also consider any non-zero $failure count
an error, and stop processing.

I will think about alternatives here, but for now it just
has to go.

Cheers,
HL

File: bin/shpec (hunk 9 above)
Comment (hlangeveld) [inline bin/shpec]: On 04/13/15 04:07, Ryland Herrick wrote:

>> +
>> +      times
>
> Didn't even know about |times|, thanks!

It's the only thing garantueed to be around.

Of course, there's the `time` keyword, but it's not a builtin
in every POSIX shell, so won't work with functions.

Even so, `times` is a work-around.  Its output is not consistent
across different shells, but at least it's there.

File: shpec/shpec_shpec.sh (hunk 10 above)
Comment (hlangeveld) [inline shpec/shpec_shpec.sh]: On 04/13/15 04:08, Ryland Herrick wrote:
> In shpec/shpec_shpec.sh <https://github.com/rylnd/shpec/pull/56#discussion_r28211219>:
>
>> @@ -35,7 +35,8 @@ describe "shpec"
>>
>>    describe "equality matcher"
>>      it "handles newlines properly"
>> -      string_with_newline_char="new\nline"
>> +      string_with_newline_char="new
>
> Was this intentionally changed? This test was added specifically for #23 <https://github.com/rylnd/shpec/issues/23>.

I've added a comment to the issue.  POSIX just sees two characters there, not a newline.

Bash and ksh support 'ansi-c' strings with the $'...' string notation.

I would recommend to use initialised string constants using `printf`.

     CHAR_NL=$( printf "\n" )

If we do this once at the top of `shpec()`, we could rewrite the assertion above as:

     string_with_newline_char="new${CHAR_NL}line"

etc., etc.

Cheers,
Henk

File: shpec/shpec_shpec.sh (hunk 11 above)
Comment (hlangeveld) [inline shpec/shpec_shpec.sh]: On 04/13/15 04:16, Ryland Herrick wrote:
> In shpec/shpec_shpec.sh <https://github.com/rylnd/shpec/pull/56#discussion_r28211373>:
>
>>      end
>>    end
>>
>>    describe "stubbing commands"
>> +    # only bash lets us redefine 'exit', so use 'false'
>
> ce61f72 <https://github.com/rylnd/shpec/commit/ce61f728abd138ae1b8e61a57fba048e4b84b15b> contains the reasoning behind this test. My
> preference is to leave that kind of thing in git, and not in the code itself.

Agreed.

> Additionally, I would say that this initial test is unnecessary; the following two tests cover both the stubbed functionality and
> the original functionality. If any part of that stub/unstub process fails, one of these commands will indicate so.
Correct. The first assertion is superfluous.

File: bin/shpec (hunk 5 above)
Comment (rylnd) [inline bin/shpec]: I'm referring to line 160. Github will always put the comment directly below the line in question.

File: bin/shpec (hunk 9 above)
Comment (rylnd) [inline bin/shpec]: :+1:

File: bin/shpec (hunk 12 above)
Comment (hlangeveld) [inline bin/shpec]: When we run the 'shpec' script,  we want to make sure that the '$0' basename matches the name we've chosen.  ( 'shpec' )

It's a variation on idiom that I have used quite a lot.  There is a long standing 
tradition to distinguish shell 'source-files' from installed scripts by appending '.sh' to the file in the source code repository.  (Going back to sccs and make in system V.)

So we need to get rid of the directory path, and we then need to match against
either "shpec" or "shpec.sh".

So, instead of the external `basename`, I use the parameter expansion modifiers:

First, `command -v $0` will return a fully-qualified path, if there is any.
Then, ${var##*/} will strip the longest left-matching string ending in "/" from $var.
Finally, ${var%.sh} will strip the string ".sh" from the right of $var, if it's there.

Short version, if the name of the script matches 'shpec' or 'shpec.sh', we call the 'shpec' function.

File: bin/shpec (hunk 5 above)
Comment (hlangeveld) [inline bin/shpec]: Line 160, I presume.  That is a left over from  a 'live comment'.

`: ` almost reads like a comment, but it's more than that.  It's actually a proper
command that accepts arguments, and discards them.

Therefore it's useful for showing the value of parameters during execution, and will
show up in '$SHELL -x script'.

It served its purpose, and should go.

File: bin/shpec (hunk 6 above)
Comment (hlangeveld) [inline bin/shpec]: Agreed

File: bin/shpec (hunk 7 above)
Comment (hlangeveld) [inline bin/shpec]: The two should be the same.  I *think* that either way should work.  I've become more confident with $( ... ) constructs, and believe that no quoting is needed in either case.

File: bin/shpec (hunk 4 above)
Comment (hlangeveld) [inline bin/shpec]: Will do.

File: bin/shpec (hunk 3 above)
Comment (hlangeveld) [inline bin/shpec]: I see. Will fix.

File: bin/shpec (hunk 2 above)
Comment (hlangeveld) [inline bin/shpec]: I don't know what I need to do to zsh to make it accept pure POSIX shell.
It's a powerful shell for interactive use, and has some nifty features, but I gave up on zsh long ago for scripting.
Sorry.

File: bin/shpec (hunk 1 above)
Comment (hlangeveld) [inline bin/shpec]: A shell script traditionally started with just ': '.  Before kernels interpreted the hasbang,
the shell would see if the first character(s) equalled ':' or ': '.  In that case it would fork itself and feed itself the script.

I don't know how I ended up adding an empty comment up there.

**Comment by hlangeveld** [comment]
On 04/10/15 18:22, Ryland Herrick wrote:
> :heart_eyes: This looks amazing. It's quite a significant PR, so please give me some time to 1) understand it and 2) provide some
> feedback.

Thanks for the feedback so far.  Most are thinks I need to fix.
Not tonight, though.

Cheers,
Henk

**Comment by hlangeveld** [comment]
On 04/13/15 04:29, Ryland Herrick wrote:
> @hlangeveld <https://github.com/hlangeveld> now that |shpec| is a function defined as a subshell command, can you speak to whether
> this affects any existing issues (for better or worse), particularly #16 <https://github.com/rylnd/shpec/issues/16> or #30
> <https://github.com/rylnd/shpec/issues/30>?

With shpec redefining itself as a function this has little effect on issue #16.

The easiest way for me is to put the code executed in the tests in a subshell and test
its exit status, or to use command substitution (which implies a subshell):

     describe "example 1 ..."
       it "explain assertion"
          (
            : subshell with code
          )
          assert matcher ...
       end
     end

     describe "example 2 ..."
       it "explain assertion"
          a=$(
            : subshell with code
          )
          b=$(
            : another subshell
          )
          assert matcher referring to "$a" and/or "$b"
       end
     end

A completely different way of passing info between the framework and a project's tests would be to make
individual tests run entirely in a subshell and report results in a fixed format, to be read by the framework.
Say a single line with status, (#pass #fail "description"), followed by multiple lines of arbitrary output
from individual assertions with more detail.

**Comment by rylnd** [comment]
Closing this in favor of the newer #58
